x-pack/plugins/lens/public/app_plugin/lens_document_equality.ts

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { isEqual } from 'lodash';
+import { FilterManager } from 'src/plugins/data/public';
+import { Document } from '../persistence/saved_object_store';
+import { DatasourceMap } from '../types';
+import { injectDocFilterReferences, removePinnedFilters } from './save_modal_container';
+
+const removeNonSerializable = (obj: Parameters<JSON['stringify']>[0]) =>
+  JSON.parse(JSON.stringify(obj));
+
+export const isLensEqual = (
+  doc1In: Document | undefined,
+  doc2In: Document | undefined,
+  injectFilterReferences: FilterManager['inject'],
+  datasourceMap: DatasourceMap
+) => {
+  if (doc1In === undefined || doc2In === undefined) {
+    return doc1In === doc2In;
+  }
+
+  // we do this so that undefined props are the same as non-existant props
+  const doc1 = removeNonSerializable(doc1In);
+  const doc2 = removeNonSerializable(doc2In);
+
+  if (doc1?.visualizationType !== doc2?.visualizationType) {
+    return false;
+  }
+
+  if (!isEqual(doc1.state.query, doc2.state.query)) {
+    return false;
+  }
+
+  if (!isEqual(doc1.state.visualization, doc2.state.visualization)) {
+    return false;
+  }
+
+  // data source equality
+  let datasourcesEqual = isEqual(
+    Object.keys(doc1.state.datasourceStates),
+    Object.keys(doc2.state.datasourceStates)
+  );
```

Comment:
@flash1293 for the purposes of the Lens equality check, does order matter in the `datasourceStates`?

Reply:
Nope, oder is irrelevant for these